File: c/linked-list/linked-list.c
#include <stdlib.h>
#include <stdio.h>



//node struct
struct Node{
    int data;
    struct Node* next;
};

//function to create nodes with data
struct Node* createNode (int data){ 

    struct Node* newNode = (struct Node*) malloc( sizeof(*newNode) );

    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}


//inserts a node at the beginning
//creates a node and inserts node at the beginning of the list 
void insertsAtBeginning(struct Node** head, int data){

    struct Node* newNode = createNode(data);

    //initialize head with  NULL
    newNode->next = *head;

    //head now points to newNode
    *head = newNode;
}


//creates a node and inserts at the end of the list
void insertAtEnd(struct Node** head, int data){

    struct Node* newNode = createNode(data);

    //head points to nothing, make newNode next
    if(*head == NULL){
        *head = newNode;
        return;
    }

    struct Node* temp = *head;
    while (temp->next != NULL){
        temp = temp->next;
    }
    temp->next = newNode;

}

//delet a node by key
void deletNode(struct Node** head, int key){

    struct Node* temp = *head;
    struct Node* prev = NULL;


    //if first after head
    if(temp != NULL && temp->data == key){
        *head = temp->next;
        free(temp);
        return;
    }

    //search for node to delete
    while(temp != NULL && temp->data != key){
        prev = temp;
        temp = temp->next;
    }

    //key was not found
    if(temp == NULL){
        return;
    }

    //remove node from the list
    prev->next = temp->next;

    free(temp);
}


///print data in the list
void printList(struct Node* head){
    struct Node* temp = head;

    while(temp != NULL){
        printf("%d -> ", temp->data);
        temp = temp->next;
    }

    printf("NULL\n");
}

File: javascript/linked-list/linked-list.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
  }

  //method to insert a node at the beginning of the list
  insertAtBeginning(data) {
    const newNode = new Node(data);

    //pointing new node to head
    //null, initially
    newNode.next = this.head;

    this.head = newNode;
  }

  //method to insert data at the end of the list
  insertAtEnd(data) {
    const newNode = new Node(data);

    //if no node exist after head node
    if (!this.head) {
      this.head = newNode;
      return;
    }

    //end of list
    let current = this.head;
    while (current.next) {
      current = current.next;
    }
    current.next = newNode;
  }

  //delet a node by value
  deleteNode(data) {
    //head points to nothing
    if (!this.head) {
      return;
    }

    //if node is the head
    if (this.head.data == data) {
      this.head = this.head.next;
      return;
    }

    let current = this.head;
    let prev = null;
    while (current && current.data !== data) {
      prev = current;
      current = current.next;
    }

    //node not found
    if (!current) {
      return;
    }

    //unlinking the node
    prev.next = current.next;
  }

  printList() {
    let current = this.head;
    let result = "";
    while (current) {
      result += current.data + "->";
      current = current.next;
    }
    console.log(result + "NULL");
  }
}
